lib/Parse/ParseDecl.cpp

```diff
@@ -3101,6 +3114,46 @@ Parser::parseDeclExtension(ParseDeclOptions Flags, DeclAttributes &Attributes) {
   return DCC.fixupParserResult(status, ext);
 }
 
+ParserResult<PoundDiagnosticDecl> Parser::parseDeclPoundDiagnostic() {
+  bool isError = Tok.is(tok::pound_error);
+  SyntaxParsingContext LocalContext(SyntaxContext, 
+    isError ? SyntaxKind::PoundErrorDecl : SyntaxKind::PoundWarningDecl);
+  SourceLoc startLoc = 
+    consumeToken(isError ? tok::pound_error : tok::pound_warning);
+
+  SourceLoc lParenLoc;
+  if (parseToken(tok::l_paren, lParenLoc, 
+                 diag::pound_diagnostic_expected, "(", isError))
+    return makeParserError();
+
+  if (!Tok.is(tok::string_literal)) {
+    diagnose(startLoc, diag::pound_diagnostic_expected_string, isError);
+    return makeParserError();
+  }
+  
+  auto string = parseExprStringLiteral();
+  if (string.isNull())
+    return makeParserError();
```

Comment:
Suggestion: continue looking for a close-paren if it occurs before the end of the line.

Reply:
👍

I originally tried checking for `#warning “Foo”` and FixIt-ing in the surrounding parens. Should I continue with that?

Reply:
That's also good, but I'm mostly thinking about `#warning(Foo)` and worse `#warning(Foo is bad, yanno?)`, where the fix is probably to add quotes.

Reply:
Also a good idea!